ngraph/core/src/op/matmul.cpp

```diff
@@ -231,10 +242,43 @@ namespace
         }
         return rc;
     }
-}
+} // namespace
 
 bool op::MatMul::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const
 {
     OV_ITT_SCOPED_TASK(itt::domains::nGraphOp, "op::MatMul::evaluate");
     return evaluate_matmul(inputs[0], inputs[1], outputs[0], get_transpose_a(), get_transpose_b());
 }
+
+void ngraph::op::v0::MatMul::validate_and_infer_types()
+{
+    element::Type result_et;
+
+    NODE_VALIDATION_CHECK(
+        this,
+        element::Type::merge(result_et, get_input_element_type(0), get_input_element_type(1)),
+        "Arguments do not have the same element type (arg0 element type: ",
+        get_input_element_type(0),
+        ", arg1 element type: ",
+        get_input_element_type(1),
+        ").");
+
+    const auto& A_matrix = get_input_partial_shape(0);
+    const auto& B_matrix = get_input_partial_shape(1);
+
+    if (A_matrix.rank().is_static() && B_matrix.rank().is_static())
+    {
+        PartialShape output_shape;
+
+        const bool transpose_a = get_transpose_a();
+        const bool transpose_b = get_transpose_b();
+
+        output_shape = validate_matmul_output_shape(A_matrix, B_matrix, transpose_a, transpose_b);
+
+        set_output_type(0, result_et, output_shape);
+    }
+    else
+    {
+        set_output_type(0, result_et, PartialShape::dynamic());
```

Comment:
Rank may have lower and upper bounds specified. In this case, you can propagate these bounds to the resulting rank. Moreover, if one of the inputs has static rank and another input has dynamic rank, depending on lower/upper bounds for the rank of the second input, the resulting rank can be static and you you may have ability to specify lower/upper bounds for each output dimension. Please estimate how many lines of code we should write to do that.

Reply:
Could you provide more details how to process lower/upper bounds for the output shape in case:
- both inputs have dynamic ranks,
- only one input has dynamic rank?

Reply:
If both have dynamic ranks, each is defined by min..max interval. Say first input have rank = [0..3], the second input have rank [0..2], then the resulting rank should be defined as interval [0..3], so upper bound is max(upper bounds on input ranks). This is a consequences of batch dimensions broadcast rules. In you current code, the output rank is completely dynamic, so upper bound is int_max that means it is undefined.

Only one dynamic. Consider MatMul({rank=0..2}, {?, 10}), then output shape has static rank=2 and even one of the dimension is defined statically: {?, 10}.